### PR TITLE
fix bug: kubelet potential panic

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
**What this PR does / why we need it**:

fix bug: kubelet may panic calling wg.Done after wg.Wait returned.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: cmd/kubelet/app/server.go [hunk 1]
```diff
@@ -946,28 +946,29 @@ func RunKubelet(kubeFlags *options.KubeletFlags, kubeCfg *kubeletconfiginternal.
...
-		}, 0, wait.NeverStop)
+		}()
 	}
 	if kubeCfg.ReadOnlyPort > 0 {
 		wg.Add(1)
-		go wait.Until(func() {
+		go func() {
 			wg.Done()
 			k.ListenAndServeReadOnly(net.ParseIP(kubeCfg.Address), uint(kubeCfg.ReadOnlyPort))
-		}, 0, wait.NeverStop)
```
Comment (hzxuzhonghu) [inline cmd/kubelet/app/server.go]: no need, because if ListenAndServe failed, kubelet will exit.

File: cmd/kubelet/app/server.go [hunk 2]
```diff
@@ -946,28 +946,29 @@ func RunKubelet(kubeFlags *options.KubeletFlags, kubeCfg *kubeletconfiginternal.
...
 	}, 0, wait.NeverStop)
 
 	// start the kubelet server
 	if enableServer {
 		wg.Add(1)
-		go wait.Until(func() {
+		go func() {
 			wg.Done()
 			k.ListenAndServe(net.ParseIP(kubeCfg.Address), uint(kubeCfg.Port), kubeDeps.TLSOptions, kubeDeps.Auth, kubeCfg.EnableDebuggingHandlers, kubeCfg.EnableContentionProfiling)
-		}, 0, wait.NeverStop)
```
Comment (hzxuzhonghu) [inline cmd/kubelet/app/server.go]: same here

File: cmd/kubelet/app/server.go [hunk 3]
```diff
@@ -946,28 +946,29 @@ func RunKubelet(kubeFlags *options.KubeletFlags, kubeCfg *kubeletconfiginternal.
 
 func startKubelet(k kubelet.Bootstrap, podCfg *config.PodConfig, kubeCfg *kubeletconfiginternal.KubeletConfiguration, kubeDeps *kubelet.Dependencies, enableServer bool) {
 	wg := sync.WaitGroup{}
+	once := sync.Once{}
 
 	// start the kubelet
 	wg.Add(1)
 	go wait.Until(func() {
-		wg.Done()
```
Comment (hzxuzhonghu) [inline cmd/kubelet/app/server.go]: panic called the second time

**Comment by hzxuzhonghu** [comment]
cc @vishh @yujuhong

File: cmd/kubelet/app/server.go [hunk 4]
```diff
@@ -946,28 +946,29 @@ func RunKubelet(kubeFlags *options.KubeletFlags, kubeCfg *kubeletconfiginternal.
...
 	if kubeCfg.ReadOnlyPort > 0 {
 		wg.Add(1)
-		go wait.Until(func() {
+		go func() {
 			wg.Done()
 			k.ListenAndServeReadOnly(net.ParseIP(kubeCfg.Address), uint(kubeCfg.ReadOnlyPort))
-		}, 0, wait.NeverStop)
+		}()
 	}
 	wg.Wait()
```
Comment (tallclair) [inline cmd/kubelet/app/server.go]: What is the point of this wait group? It looks like it's waiting for the goroutines to start, but why? It's not actually checking that the servers in those go routines have started...

Comment (hzxuzhonghu) [reply cmd/kubelet/app/server.go]: If listen and serve failed , process will exit.
```
// ListenAndServeKubeletReadOnlyServer initializes a server to respond to HTTP network requests on the Kubelet.
func ListenAndServeKubeletReadOnlyServer(host HostInterface, resourceAnalyzer stats.ResourceAnalyzer, address net.IP, port uint, runtime kubecontainer.Runtime) {
	glog.V(1).Infof("Starting to listen read-only on %s:%d", address, port)
	s := NewServer(host, resourceAnalyzer, nil, false, false, runtime, nil)

	server := &http.Server{
		Addr:           net.JoinHostPort(address.String(), strconv.FormatUint(uint64(port), 10)),
		Handler:        &s,
		MaxHeaderBytes: 1 << 20,
	}
	glog.Fatal(server.ListenAndServe())
}
```

Comment (tallclair) [reply cmd/kubelet/app/server.go]: I don't think that answers my question. Why can't you just remove the waitgroup entirely? It doesn't appear to be doing anything useful here.

Comment (hzxuzhonghu) [reply cmd/kubelet/app/server.go]: The wa.Add（） after wg.Wait may cause a panic.

Comment (hzxuzhonghu) [reply cmd/kubelet/app/server.go]: And I thought wg wait for server really startup.

Comment (hzxuzhonghu) [reply cmd/kubelet/app/server.go]: I should remove the waitgroup.

**Comment by hzxuzhonghu** [comment]
@tallclair Done

**Comment by tallclair** [comment]
/lgtm

thanks!

**Comment by hzxuzhonghu** [comment]
@dchen1107 for approval

**Comment by hzxuzhonghu** [comment]
kindly ping @derekwaynecarr for approval

**Comment by hzxuzhonghu** [comment]
@kubernetes/sig-node-pr-reviews  for approval

**Pull request merged**